src/tfr_sensor/src/light_detection_action_server.cpp

```diff
@@ -0,0 +1,90 @@
+#include <ros/ros.h>
+#include <ros/console.h>
+#include <light_detector.h>
+#include <tfr_msgs/LightDetectAction.h>
+#include <actionlib/server/simple_action_server.h>
+#include <image_transport/image_transport.h>
+
+/*
+ *  An action server that looks for the switching on of a light.  
+ *  Action message is empty,it merely signals the server to start processing.
+ *  
+ *  The server will not examine anything until commanded, and will set it's
+ *  status to succeeded, when it sees the light.
+ * */
+class DetectionActionServer
+{
+    public:
+        DetectionActionServer(ros::NodeHandle &node, const std::string name,
+                const std::string camera_name, int window_size,
+                double threshold) : 
+            n{node},
+            server{node, name, false},
+            detector{window_size, threshold},
+            it{node}
+        {
+            server.registerGoalCallback(
+                    boost::bind(&DetectionActionServer::setGoal,this));
+            server.registerPreemptCallback(
+                    boost::bind(&DetectionActionServer::preempt,this));
+            image_subscriber = it.subscribe(camera_name, 5,
+                    &DetectionActionServer::detect, this);
+            ROS_INFO("DetectionActionServer starting connection");
+            server.start();
+            ROS_INFO("DetectionActionServer connected");
+        }
+
+        ~DetectionActionServer() = default;
+        DetectionActionServer( const DetectionActionServer&) = delete;
+        DetectionActionServer& operator=(const DetectionActionServer&) = delete;
+        DetectionActionServer(DetectionActionServer&&) = delete;
+        DetectionActionServer& operator=(DetectionActionServer&&)=delete;
+        
+    private:
+        void setGoal()
+        {
+            ROS_INFO("DetectionActionServer accepted goal");
+            detector.clear();
+            server.acceptNewGoal();
+        }
+        void preempt()
+        {
+            ROS_INFO("DetectionActionServer preempted goal");
+            server.setPreempted();
+        }
+        void detect(const sensor_msgs::ImageConstPtr& msg)
+        {
+
+            if (!server.isActive() || !ros::ok())
+                return;
+            detector.add_image(msg);
+            if (detector.is_on())
+                server.setSucceeded();
+        }
+
+        ros::NodeHandle &n;
+        actionlib::SimpleActionServer<tfr_msgs::LightDetectAction> server;
+        LightDetector detector;
+        image_transport::ImageTransport it;
```

Comment:
I'd like a longer name for this variable.

Reply:
I can do that.